File: swift/Sources/CircleAI/Catalog.swift
// Catalog.swift
//
// ModelScopeCatalogClient + signature verifier + ModelEntry / ModelRegistry.

import Foundation
#if canImport(FoundationNetworking)
import FoundationNetworking
#endif

public enum CatalogSignatureResult: Int, Sendable {
    case valid = 0
    case invalid = 1
    case missing = 2
    case notConfigured = 3
}

public protocol ICatalogSignatureVerifier: Sendable {
    func verify(payload: Data, signatureBase64: String?) -> CatalogSignatureResult
}

public struct NullCatalogSignatureVerifier: ICatalogSignatureVerifier {
    public init() {}
    public func verify(payload: Data, signatureBase64: String?) -> CatalogSignatureResult { .notConfigured }
}

public enum CatalogRefreshCadence: Int, Sendable {
    case onStartup = 0
    case daily = 1
    case manual = 2
    case never = 3
}

public struct ModelScopeCatalogOptions: Sendable {
    public let baseUri: String
    public let cacheDirectory: String
    public let cadence: CatalogRefreshCadence
    public let filter: String
    public let pageSize: Int
    public let userAgent: String

    public init(
        baseUri: String = "https://www.modelscope.cn",
        cacheDirectory: String = Self.defaultCacheDir(),
        cadence: CatalogRefreshCadence = .onStartup,
        filter: String = "MNN",
        pageSize: Int = 100,
        userAgent: String = "Mozilla/5.0 (Circle AI SDK) CircleAI-Swift/1.5"
    ) {
        self.baseUri = baseUri
        self.cacheDirectory = cacheDirectory
        self.cadence = cadence
        self.filter = filter
        self.pageSize = pageSize
        self.userAgent = userAgent
    }

    public static func defaultCacheDir() -> String {
        let home = FileManager.default.homeDirectoryForCurrentUser.path
        return (home as NSString).appendingPathComponent(".circleai/catalog")
    }
}

public struct ModelEntry: Codable, Sendable, Equatable {
    public let name: String
    public let version: String
    public let quantization: String
    public let url: String?
    public let checksum: String?
    public let repo: String?
    public let totalBytes: Int64
    public let bundleFiles: [BundleFile]
    public let minRamGb: Double
    public let minStorageGb: Double
    public let capabilities: [String]?
    public let qualityRank: Int

    public init(
        name: String,
        version: String,
        quantization: String = "",
        url: String? = nil,
        checksum: String? = nil,
        repo: String? = nil,
        totalBytes: Int64 = 0,
        bundleFiles: [BundleFile] = [],
        minRamGb: Double = 0,
        minStorageGb: Double = 0,
        capabilities: [String]? = nil,
        qualityRank: Int = 0
    ) {
        self.name = name; self.version = version; self.quantization = quantization
        self.url = url; self.checksum = checksum; self.repo = repo
        self.totalBytes = totalBytes; self.bundleFiles = bundleFiles
        self.minRamGb = minRamGb; self.minStorageGb = minStorageGb
        self.capabilities = capabilities; self.qualityRank = qualityRank
    }

    public var isBundle: Bool { !bundleFiles.isEmpty }

    enum CodingKeys: String, CodingKey {
        case name, version, quantization, url, checksum, repo
        case totalBytes = "total_bytes"
        case bundleFiles = "bundle_files"
        case minRamGb = "min_ram_gb"
        case minStorageGb = "min_storage_gb"
        case capabilities
        case qualityRank = "quality_rank"
    }
}

public struct ModelRegistry: Codable, Sendable {
    public let registryUrl: String
    public let lastUpdated: Date
    public let models: [ModelEntry]

    public init(registryUrl: String, lastUpdated: Date, models: [ModelEntry]) {
        self.registryUrl = registryUrl; self.lastUpdated = lastUpdated; self.models = models
    }

    enum CodingKeys: String, CodingKey {
        case registryUrl = "registry_url"
        case lastUpdated = "last_updated"
        case models
    }
}

public actor ModelScopeCatalogClient {
    private let options: ModelScopeCatalogOptions
    private let verifier: ICatalogSignatureVerifier
    private let networkTypeProvider: (@Sendable () -> String?)?
    private var refreshedThisRun: Bool = false
    private let session: URLSession

    public init(
        options: ModelScopeCatalogOptions = ModelScopeCatalogOptions(),
        verifier: ICatalogSignatureVerifier = NullCatalogSignatureVerifier(),
        networkTypeProvider: (@Sendable () -> String?)? = nil
    ) {
        self.options = options
        self.verifier = verifier
        self.networkTypeProvider = networkTypeProvider
        let cfg = URLSessionConfiguration.ephemeral
        cfg.timeoutIntervalForRequest = 10
        self.session = URLSession(configuration: cfg)
        try? FileManager.default.createDirectory(atPath: options.cacheDirectory, withIntermediateDirectories: true)
    }

    public nonisolated var cacheFilePath: String {
        (options.cacheDirectory as NSString).appendingPathComponent("catalog.json")
    }
    public nonisolated var signatureFilePath: String {
        (options.cacheDirectory as NSString).appendingPathComponent("catalog.sig")
    }

    public func isRefreshDue() async -> Bool {
        if options.cadence == .never || options.cadence == .manual { return false }
        if let prov = networkTypeProvider {
            if let net = prov()?.lowercased(), net == "none" { return false }
        }
        if !FileManager.default.fileExists(atPath: cacheFilePath) { return true }
        if options.cadence == .onStartup { return !refreshedThisRun }
        // Daily.
        guard let attrs = try? FileManager.default.attributesOfItem(atPath: cacheFilePath),
              let mtime = attrs[.modificationDate] as? Date else { return false }
        let cal = Calendar(identifier: .gregorian)
        return !cal.isDate(mtime, inSameDayAs: Date())
    }

    public nonisolated func loadFromDisk() -> ModelRegistry? {
        guard let data = try? Data(contentsOf: URL(fileURLWithPath: cacheFilePath)) else { return nil }
        let decoder = JSONDecoder()
        decoder.dateDecodingStrategy = .iso8601
        return try? decoder.decode(ModelRegistry.self, from: data)
    }

    public func getCachedCatalog(acceptStaleOnError: Bool = true) async throws -> ModelRegistry? {
        if await isRefreshDue() {
            do {
                return try await refresh()
            } catch {
                if !acceptStaleOnError { throw error }
            }
        }
        return loadFromDisk()
    }

    public func refresh() async throws -> ModelRegistry {
        let reg = try await fetchLive()
        let encoder = JSONEncoder()
        encoder.dateEncodingStrategy = .iso8601
        encoder.outputFormatting = .prettyPrinted
        let bytes = try encoder.encode(reg)

        var existingSig: String?
        if let data = try? Data(contentsOf: URL(fileURLWithPath: signatureFilePath)),
           let s = String(data: data, encoding: .utf8) {
            existingSig = s.trimmingCharacters(in: .whitespacesAndNewlines)
        }
        let sigResult = verifier.verify(payload: bytes, signatureBase64: existingSig)
        if sigResult == .invalid {
            throw NSError(domain: "CircleAI.Catalog", code: 1, userInfo: [
                NSLocalizedDescriptionKey: "Catalog signature did not verify."
            ])
        }
        try? FileManager.default.createDirectory(atPath: options.cacheDirectory, withIntermediateDirectories: true)
        try bytes.write(to: URL(fileURLWithPath: cacheFilePath))
        refreshedThisRun = true
        return reg
    }

    private func fetchLive() async throws -> ModelRegistry {
        var comps = URLComponents(string: "\(options.baseUri)/api/v1/models")!
        comps.queryItems = [
            URLQueryItem(name: "Name", value: options.filter),
            URLQueryItem(name: "PageSize", value: String(options.pageSize)),
        ]
        let listing = try await fetchJson(comps.url!)
        let dataNode = listing["Data"] as? [String: Any] ?? [:]
        let items = dataNode["Model"] as? [[String: Any]] ?? []

        var entries: [ModelEntry] = []
        for m in items {
            guard let name = m["Name"] as? String, let path = m["Path"] as? String,
                  !name.isEmpty, !path.isEmpty else { continue }
            let filesURL = URL(string: "\(options.baseUri)/api/v1/models/\(path)/repo/files?Revision=master")!
            let filesJson: [String: Any]
            do { filesJson = try await fetchJson(filesURL) } catch { continue }
            let filesData = filesJson["Data"] as? [String: Any] ?? [:]
            let files = filesData["Files"] as? [[String: Any]] ?? []
            var bundle: [BundleFile] = []
            var total: Int64 = 0
            for f in files {
                let n = (f["Path"] as? String) ?? (f["Name"] as? String) ?? ""
                if n.isEmpty { continue }
                let size = (f["Size"] as? NSNumber)?.int64Value ?? 0
                bundle.append(BundleFile(name: n, sha256: (f["Sha256"] as? String) ?? "", sizeBytes: size))
                total += size
            }
            entries.append(ModelEntry(
                name: name,
                version: (m["Revision"] as? String) ?? "master",
                quantization: (m["Quantization"] as? String) ?? "",
                repo: path,
                totalBytes: total,
                bundleFiles: bundle
            ))
        }
        return ModelRegistry(registryUrl: options.baseUri, lastUpdated: Date(), models: entries)
    }

    private func fetchJson(_ url: URL) async throws -> [String: Any] {
        var req = URLRequest(url: url)
        req.setValue(options.userAgent, forHTTPHeaderField: "User-Agent")
        let (data, resp) = try await session.data(for: req)
        if let http = resp as? HTTPURLResponse, http.statusCode != 200 {
            throw NSError(domain: "CircleAI.Catalog", code: http.statusCode)
        }
        return (try JSONSerialization.jsonObject(with: data) as? [String: Any]) ?? [:]
    }
}

File: android/circle-ai/src/main/kotlin/com/bhengubv/circleai/android/catalog/Catalog.kt
// Catalog.kt
//
// ModelEntry + ModelRegistry + ModelScopeCatalogClient + signature verifier.
// Port of CircleAI.Core.Models.ModelScopeCatalogClient.

package com.bhengubv.circleai.android.catalog

import com.bhengubv.circleai.android.models.BundleFile
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import java.io.BufferedReader
import java.io.File
import java.io.InputStreamReader
import java.net.HttpURLConnection
import java.net.URL
import java.net.URLEncoder
import java.nio.charset.StandardCharsets
import java.time.Instant
import java.time.LocalDate
import java.time.ZoneOffset

enum class CatalogSignatureResult { VALID, INVALID, MISSING, NOT_CONFIGURED }

interface ICatalogSignatureVerifier {
    fun verify(payload: ByteArray, signatureBase64: String?): CatalogSignatureResult
}

object NullCatalogSignatureVerifier : ICatalogSignatureVerifier {
    override fun verify(payload: ByteArray, signatureBase64: String?): CatalogSignatureResult =
        CatalogSignatureResult.NOT_CONFIGURED
}

enum class CatalogRefreshCadence { ON_STARTUP, DAILY, MANUAL, NEVER }

data class ModelScopeCatalogOptions(
    val baseUri: String = "https://www.modelscope.cn",
    val cacheDirectory: String = defaultCacheDir(),
    val cadence: CatalogRefreshCadence = CatalogRefreshCadence.ON_STARTUP,
    val filter: String = "MNN",
    val pageSize: Int = 100,
    val userAgent: String = "Mozilla/5.0 (Circle AI SDK) CircleAI-Kotlin/1.5",
) {
    companion object {
        fun defaultCacheDir(): String =
            File(System.getProperty("user.home"), ".circleai/catalog").path
    }
}

data class ModelEntry(
    val name: String,
    val version: String,
    val quantization: String = "",
    val url: String? = null,
    val checksum: String? = null,
    val repo: String? = null,
    val totalBytes: Long = 0,
    val bundleFiles: List<BundleFile> = emptyList(),
    val minRamGb: Double = 0.0,
    val minStorageGb: Double = 0.0,
    val capabilities: List<String>? = null,
    val qualityRank: Int = 0,
) {
    val isBundle: Boolean get() = bundleFiles.isNotEmpty()
}

data class ModelRegistry(
    val registryUrl: String,
    val lastUpdated: Instant,
    val models: List<ModelEntry>,
)

class ModelScopeCatalogClient(
    private val options: ModelScopeCatalogOptions = ModelScopeCatalogOptions(),
    private val verifier: ICatalogSignatureVerifier = NullCatalogSignatureVerifier,
    private val networkTypeProvider: (() -> String?)? = null,
) {
    private var refreshedThisRun: Boolean = false

    val cacheFilePath: String get() = File(options.cacheDirectory, "catalog.json").path
    val signatureFilePath: String get() = File(options.cacheDirectory, "catalog.sig").path

    init {
        File(options.cacheDirectory).mkdirs()
    }

    suspend fun isRefreshDueAsync(): Boolean = withContext(Dispatchers.IO) {
        if (options.cadence == CatalogRefreshCadence.NEVER) return@withContext false
        if (options.cadence == CatalogRefreshCadence.MANUAL) return@withContext false
        networkTypeProvider?.let {
            val net = it()?.lowercase()
            if (net == "none") return@withContext false
        }
        val cacheFile = File(cacheFilePath)
        if (!cacheFile.exists()) return@withContext true
        if (options.cadence == CatalogRefreshCadence.ON_STARTUP) {
            return@withContext !refreshedThisRun
        }
        val mtimeDate = Instant.ofEpochMilli(cacheFile.lastModified())
            .atOffset(ZoneOffset.UTC).toLocalDate()
        mtimeDate < LocalDate.now(ZoneOffset.UTC)
    }

    fun loadFromDisk(): ModelRegistry? {
        return try {
            val file = File(cacheFilePath)
            if (!file.exists()) return null
            parseRegistryJson(file.readText(StandardCharsets.UTF_8))
        } catch (_: Throwable) {
            null
        }
    }

    suspend fun getCachedCatalogAsync(acceptStaleOnError: Boolean = true): ModelRegistry? {
        return try {
            if (isRefreshDueAsync()) refreshAsync() else loadFromDisk()
        } catch (e: Throwable) {
            if (acceptStaleOnError) loadFromDisk() else throw e
        }
    }

    suspend fun refreshAsync(): ModelRegistry = withContext(Dispatchers.IO) {
        val reg = fetchLive()
        val json = registryToJson(reg)
        val bytes = json.toByteArray(StandardCharsets.UTF_8)

        val sigFile = File(signatureFilePath)
        val existingSig = if (sigFile.exists()) sigFile.readText().trim().ifBlank { null } else null

        val sigResult = verifier.verify(bytes, existingSig)
        if (sigResult == CatalogSignatureResult.INVALID) {
            throw RuntimeException("Catalog signature did not verify against the configured public key.")
        }

        File(options.cacheDirectory).mkdirs()
        File(cacheFilePath).writeBytes(bytes)
        refreshedThisRun = true
        reg
    }

    // ── Network ─────────────────────────────────────────────────────────

    private fun fetchLive(): ModelRegistry {
        val listingUrl = "${options.baseUri}/api/v1/models?Name=" +
            URLEncoder.encode(options.filter, StandardCharsets.UTF_8) +
            "&PageSize=${options.pageSize}"
        val listing = httpGetJson(listingUrl)
        val items = extractListingItems(listing)
        val entries = mutableListOf<ModelEntry>()
        for (item in items) {
            val name = item["Name"] ?: continue
            val repoPath = item["Path"] ?: continue
            val filesUrl = "${options.baseUri}/api/v1/models/$repoPath/repo/files?Revision=master"
            val fileBody = try { httpGetJson(filesUrl) } catch (_: Throwable) { continue }
            val files = extractFileItems(fileBody)
            var total = 0L
            for (f in files) total += f.sizeBytes
            entries.add(
                ModelEntry(
                    name = name,
                    version = item["Revision"] ?: "master",
                    quantization = item["Quantization"] ?: "",
                    repo = repoPath,
                    totalBytes = total,
                    bundleFiles = files,
                )
            )
        }
        return ModelRegistry(options.baseUri, Instant.now(), entries)
    }

    private fun httpGetJson(url: String): String {
        val conn = URL(url).openConnection() as HttpURLConnection
        try {
            conn.requestMethod = "GET"
            conn.setRequestProperty("User-Agent", options.userAgent)
            conn.connectTimeout = 10_000
            conn.readTimeout = 30_000
            val code = conn.responseCode
            if (code != 200) throw RuntimeException("HTTP $code fetching $url")
            BufferedReader(InputStreamReader(conn.inputStream, StandardCharsets.UTF_8)).use { r ->
                return r.readText()
            }
        } finally {
            conn.disconnect()
        }
    }
}
